dimagi/utils/django/management.py

```diff
@@ -11,3 +16,40 @@ def are_you_sure(prompt="Are you sure you want to proceed? (yes or no): "):
     except Exception:
         return False
 
+
+def export_as_csv_action(description="Export selected objects as CSV file",
+                         fields=None, exclude=None, header=True):
+    """
+    Include this in ModelAdmin.actions to get an option of export as CSV in Django Admin.
+
+    copied from https://djangosnippets.org/snippets/2369/
+
+    This function returns an export csv action
+    'fields' and 'exclude' work like in django ModelForm
+    'header' is whether or not to output the column names as the first row
+    """
+    def export_as_csv(modeladmin, request, queryset):
+        """
+        Generic csv export admin action.
+        based on http://djangosnippets.org/snippets/1697/
+        """
+        opts = modeladmin.model._meta
+        field_names = set([field.name for field in opts.fields])
+        if fields:
+            fieldset = set(fields)
+            field_names = field_names & fieldset
+        elif exclude:
+            excludeset = set(exclude)
+            field_names = field_names - excludeset
+
+        response = HttpResponse(mimetype='text/csv')
+        response['Content-Disposition'] = 'attachment; filename=%s.csv' % unicode(opts).replace('.', '_')
+
+        writer = csv.writer(response)
+        if header:
+            writer.writerow(list(field_names))
+        for obj in queryset:
+            writer.writerow([unicode(getattr(obj, field)).encode("utf-8", "replace") for field in field_names])
```

Comment:
`.encode("utf-8", "replace")` doesn't make much sense. UTF-8 can encode any Unicode character.

The place where you need to worry is `unicode(getattr(obj, field))`, which could fail with a `UnicodeDecodeError` if `getattr(obj, field)` returns a byte string since it will try to decode it using the default (ASCII) encoding. To fix that issue you can use `unicode(..., 'utf-8', 'replace')`. However, that will fail with a `TypeError` if `getattr(...)` returns a `unicode` object. To do this safely you need a conditional:

```py
def safe_utf8(value):
    if isinstance(value, bytes):
        # This will needlessly decode, and then re-encode UTF-8 strings.
        # All other encodings will be converted to UTF-8 (the conversion
        # is potentially lossy).
        return value.decode('utf-8', 'replace').encode('utf-8')
    return unicode(value).encode('utf-8')
```
This function should work in Python 2.6+